### PR TITLE
[MXNET-851] Test coverage metrics for R-package

### PR DESCRIPTION
**Pull request by anirudhacharya**
## Description ##
Track test coverage metrics for R-package using `covr` package.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- @marcoabreu

### CODE REVIEW

**Comment by marcoabreu** [comment]
Thank you! Could you elaborate how and when this script is getting executed? I guess we have to modify our test pipeline and add it there

**Comment by marcoabreu** [comment]
https://github.com/apache/incubator-mxnet/blob/master/Makefile#L591

**Comment by marcoabreu** [comment]
https://github.com/apache/incubator-mxnet/blob/master/ci/docker/runtime_functions.sh#L738

**Comment by marcoabreu** [comment]
Thanks. Did you enable the proper error propagation?

**Comment by marcoabreu** [comment]
Great, thank you.

Do you want to fix that in a separate PR?

File: R-package/ci-test-coverage.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ # "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ # KIND, either express or implied.  See the License for the
+ # specific language governing permissions and limitations
+ # under the License.
+ 
+set -evx
+  
+MXNET_HOME=${PWD}
+cd ${MXNET_HOME}/R-package
+Rscript -e 'covr::codecov()'
```
Comment (marcoabreu) [inline R-package/ci-test-coverage.sh]: I think this script is obsolete now

**Comment by anirudhacharya** [comment]
test coverage report - http://jenkins.mxnet-ci.amazon-ml.com/blue/rest/organizations/jenkins/pipelines/incubator-mxnet/branches/PR-12391/runs/6/nodes/133/steps/154/log/?start=0

@marcoabreu The CI run generates the report but it is not able to upload the reports to codecov due to the following error - 
```
Please provide the repository token to upload reports via `-t :repository-token`
```

**Comment by marcoabreu** [comment]
Do you see the publish_test_coverage call in our Jenkinsfile? It's important that this call submits the report to codecov. Here, it seems like the package is trying to upload the report itself. Is it possible to just output the test coverage report into the workdir? Codecov will then automatically pick it up

**Comment by marcoabreu** [comment]
Have a look at the source here: https://github.com/r-lib/covr/blob/master/R/codecov.R

They automatically try to upload to codecov. We don't want that. Instead, we want this output to be redirected to a file. Our method then automatically picks that file up and uploads it properly.

Basically what I would recommend is to take that function and replicate it to redirect the output to a file. Can you try that?

The reason we don't want this implicit upload is because of high availability. We took some measures because codecov is not really reliable in terms of availability of their default endpoints.

**Comment by anirudhacharya** [comment]
@marcoabreu please review.

**Comment by marcoabreu** [comment]
Can you upload the content of that file on pastebin or gist? I'd love to have a closer look to investigate

**Comment by anirudhacharya** [comment]
an example test coverage json - https://drive.google.com/file/d/1xlynn-rBtpEPVl6De64fB76qjeO-Nh__/view?usp=sharing

File: Makefile [hunk 2]
```diff
@@ -579,16 +578,15 @@ rpkg:
...
 	devtools::install_version('roxygen2',version='5.0.1',\
 	repo='https://cloud.r-project.org/',quiet=TRUE)}"
 	Rscript -e "require(roxygen2); roxygen2::roxygenise('R-package')"
 	R CMD INSTALL R-package
-	rm -rf R-package/src/image_recordio.h
 
 rpkgtest:
-	Rscript -e "require(testthat);res<-test_dir('R-package/tests/testthat');if(!testthat:::all_passed(res)){stop('Test failures', call. = FALSE)}"
+	Rscript -e 'require(testthat);res<-test_dir("R-package/tests/testthat");if(!testthat:::all_passed(res)){stop("Test failures", call. = FALSE)}'
+	Rscript -e 'res<-covr:::package_coverage("R-package");fileConn<-file("r-package_coverage.json");writeLines(RJSONIO:::toJSON(res), fileConn);close(fileConn)'
```
Comment (marcoabreu) [inline Makefile]: I think this functions output does not really match the format codecov expects. Have a look at this line:

https://github.com/r-lib/covr/blob/master/R/codecov.R#L164

**Comment by marcoabreu** [comment]
Thanks a lot! It seems like codecov is having trouble associating the proper commit hash of a PR with the report, but that's a separate problem. This PR has successfully added test coverage for R, great job!

**Comment by anirudhacharya** [comment]
@marcoabreu Thanks! Though the current coverage report is not completely accurate, it shows 0% coverage for certain files which have unit tests for them. Will check why `covr` misses the coverage in those files.

**Pull request merged**